Implement M4->M0 message queue

MalcolmRobb opened the pull request:
Remove the current phase code.

Implement a queue so that M4 can post messages into the queue and M0 can
pull the messages out. The messages encode a buffer offset and length.
This allows M4 to tell M0 to transmit X bytes from offset Y over the USB
bus.

The advantage of this is that the M0 code can be common for all
byte/word packing carried out by M4. Also up to 3 commands can be queued
allowing M0 to catch up if it is slow to transmit over the USB.

MalcolmRobb committed:
Implement M4->M0 message queue

Remove the current phase code.

Implement a queue so that M0 can post messages into the queue and M0 can
pull the messages out. The messages encode a buffer offset and length.
This allows M4 to tell M0 to transmit X bytes from offset Y over the USB
bus.

The advantage of this is that the M0 code can be common for all
byte/word packing carried out by M4. Also up to 3 commands can be queued
allowing M0 to catch up if it is slow to transmit over the USB.

bvernoux commented:
Nice feature thanks,
It will need some tests to check if performance are same or better as before especially at 10MSPS.
I will do a new test trunk for new feature/refactoring in order to avoid any potential problems before to push it to trunk (which is intended to be stable and tested).

MalcolmRobb commented:
Ok - I've done some performance monitoring - and the news isn't good.

I stuck some extra code in the M4_isr to pulse a GPIO line if the ADCHS FIFO_OVERFLOW bit gets set whilst running. This will add a few cycles to the isr run time, but shouldn't be too significant. I then monitored the line on a scope running at 20MSPS.

Using the non-packing original release code, the line pulses occasionally - perhaps 10% of the time.
Using the current packing release code, the line pulses about 50% of the time. 
Using this pull, non packing, the line pulses about 75% of the time.
Using this pull, with packing, the line pulses about 75% of the time.

If I reduce the sample rate to 10MSPS, then all the pulses stop.

I then commented out all the signal_sev() statements and tried again. Doing this means the M4 is still running doing the packing, but it never kicks M0 so there isn't any USB output. In this mode I can run the ADC at 40MSPS and I never see any pulses.

So it seems to me that there are no contention problems between ADCHS/DMA and M4. However, contention between ADCHS/DMA and USB is a big problem, and that problem exists even in the 20MSPS original release code.

I see there are comments in the linker scripts about the USB using two separate 16Kb blocks of memory, and the original code alternately fills these two blocks in a Ping-Pong fashion. However, the packing code divides these two blocks in half again, and treats them as 4 equal 8Kb blocks. The problem *may* be that this causes the ADC to be DMA'ing into the same block as the USB is reading out of - hence a conflict and reduced performance.

I can't see any documentation about how the USB reads data out of the memory - does it use DMA? and if it does use DMA what is the block size? If the USB-DMA blocks the ADC-DMA for more than 16 samples, then the ADC FIFO will overflow and data will be lost - which is what I'm seeing.

bvernoux commented:
USB have it's own internal DMA controller and it is why I have separated those access using 2 separated bank of 16KB (with flip/flop) to avoid bus contention.
About the usage of 4 buffers of 8KB (used for packing) it is clear sometimes USB & ADCHS use the same bank .. and yes ADCHS FIFO is only 2*8bytes so ADCHS overflow happen in that case and we loose some samples.
I have tested that during months of development on this driver/concept in order to obtain 0 ADCHS FIFO overflow (at 10MSPS when the PC have a good USB port as if USB on PC side is too slow that can happen too) and the best was using 2 x 16KB with separated banks and also enable the statistics to check no error happens on ADCHS side (as when something go wrong like PC USB which is too slow to retrieve samples there is some errors in those counters mainly ADCHS fifo overflow).

When you say your pull pulse 75% of time have you some accurate measurement in uS or less to compare ? (before / after)
It is why I'm a bit paranoid to any new code pushed ;-) because a little things can have very bad effect.

MalcolmRobb commented:
Accurate measurements - no. And I fully understand your caution on new pull requests into the firmware.

What I did was pulse a GPIO pin in the M4_main() function every time it comes out of the signal_wfe(). So I get a steady stream of pulses at 409.6us  (for 16K blocks) or 204.8 us (for 8K blocks). 

Then inside the M4_isr, I read the LPC_ADCHS->STATUS0 register and toggle another GPIO pin if FIFO_OVERFLOW is set (and clear LPC_ADCHS->STATUS0).

So I only get overflow indication in the interrupt routine - and I can only tell on a per-block basis if there are any overflows - not how many there are in each block. 

What I see is on the original non packing code there are occasional toggles - from what you say these may be due to my test PC which does struggle to keep up at 20MSPS.   

However, there are many more toggles when running the packing code, and even more when running this pull. This can't be due to my PC because there shouldn't be any difference to the normal 20MSPS rates. I've done quite a lot of testing, and convinced myself that the problem is inside the Airspy, but it's not contention between the ADC-DMA and either the M0 or M4 processors - which means that the packing code isn't the cause. However, the USB block does appear to stall the ADC-DMA, so changing the way the memory blocks are allocated does have a big affect.

I suppose the next test is to revert to the original 2x16Kb buffers, and run the packing code on that. If the error rates stay the same then that would be a better solution than the current 4x8Kb buffers.

I also need to read up on the USB DMA and how that works..

MalcolmRobb commented:
More experiments, more observations.

The packing code seems to generate errors almost exactly 50% of the time. It seems highly likely that this occurs when the ADC-DMA is writing new data into the same block of memory that the USB-DMA is reading from. Obviously this happens 50% of the time with the current release packing code.

To test this theory, I modified the code so that the ADC-DMA only wrote to the memory from 0x20004000 to 0x20003FFF, and the USB-DMA only reads from 0x20008000-0x200BFFF. I setup the DMA to interrupt every 0x2000 bytes (so 2 times per block), and used M4 to manually copy (or pack) the data between the input ADC and output USB buffers. 

The results : There were no ADC overflow errors, with either a direct copy, or when packing between the two buffers. Infact, I could even sample at 24Mhz, and using the packing routines generate an 18Word/sec sample stream. This seems to prove that there are no contention problems between ADC-DMA and M4, or between M4 and USB-DMA. 

It uses about 45% M4 CPU to copy data from ADC to USB buffers  (@20MSPS, 140Mhz CPU)
It uses about 75% M4 CPU to pack data from ADC to USB buffers  (@20MSPS, 140Mhz CPU)
It uses about 95% M4 CPU to pack data from ADC to USB buffers  (@24MSPS, 140Mhz CPU)

Code optimisation and/or writing in assembler may shave a few tens of percent off these figures if we're lucky. And obviously we can easily increase the CPU clock to 180Mhz or 200Mhz if more grunt it required. 

A further observation is that on my PC, I cannot sustain 20MSPS (raw unpacked) unless the USB output buffers are at least 0x2000 bytes long. I split the ADC and USB buffers into 4 chunks each 0x1000 bytes long, and got the DMA to interrupt every 0x1000 bytes. I had no problems with the ADC, or copying the data from ADC to USB buffers. However, the USB link would only maintain about 19.75 MSPS.  Increasing the ADC and USB buffers back to 0x2000 bytes each allows sustained 20 MSPS operation. This may be a limitation on my PC, but, since larger USB output buffers appear to allow higher sustained USB transfer rates it appears to me that performance gains may be available by using trading smaller ADC-DMA buffers for larger USB-DMA ones.

My next tests are to Byte pack 32MSPS and 40MSPS ADC samples and see if there are any ADC overflows - I suspect there won't be.

MalcolmRobb commented:
I've now tested 32MSPS and 40MSPS, and all seems well when 8 bit byte packing - no ADC overflows, and USB data rates of 16/20 Mb/sec. 

I've coded the byte packing in assembler insert, but I can't optimize it enough to get it to work at 40MSPS with the CPU clock set to  (the default) 140Mhz. However, increasing the CPU clock to 160Mhz does allow 8 bit packed @40MSPS.

bvernoux commented:
Nice update thanks.
The other things to check later is if 40MHz is clean.

MalcolmRobb committed:
Add support for user selectable packing

Change the DMA structures so that the ADCHS DMA uses a buffer located in
one memory block, and the USB DMA uses a different memory block. Then
use the M4 CPU core to manually move the data between the ADCHS DMA
memory and the USB DMA memory. This avoids the DMA contention that
occurrs whenever the USB and ADCHS DMA accesses attempts to use the same
memory block.

Add initial support for user selectable ADC sample packing via static
variable "airspy_format". Currently implemented are 16, 12 and 8 bit
packing - but other options could be implemented (decimation, companding
etc). Included are both C and inline assembler versions of the same
routines. Currently the output format is fixed at 16 bit
(AIRSPY_FORMAT_PACK16) awaiting implementation of an agreed USB
interface way of selecting the desired output format(s).

MalcolmRobb committed:
BUGFIX : PR02

Incorrect merging of different branches. Sorry.
